# MSD
Work on this project was done using the high performance computer cluster (HPC) at the University of Essex. 

## Pre-processing
For each step of the pre-processing, there are two scripts named after the tool used (e.g. fastqc). One R script (e.g. <i>fastqc.r</i>) which generates a script with bash commands and a bash wrapper script (e.g. <i>fastqcWrapper.sh</i>) that submits the script previously generated by the R script via qsub to the HPC. Only the R and wrapper scripts are included on here. 

For pre-processing steps where the type of reads (i.e. single end or pair end) was relevant (e.g. aligning to the genome), separate scripts were created and are denoted by SE and PE in the file name for single end and pair end respectively. The pre-processing scripts were used in the following order:

1. cat_files_SE/PE.r - to merge experiment replicates 
2. fastqc.r / fastqcWrapper.sh - for quality control
3. trimmomaticSE/PR.r / trimmommaticWrapperSE/PE.sh - to trim adapters sequences
4. bowtie_SE/PE.r / bowtieWrapperSE/PE.sh - to align the data to the genome
5. sam2bamSE.r / sam2bamSEcmds.sh / sam2bamSEWrapper.sh - to convert the sam files to bam - this was optional and was not performed for all dada due to some issues with peak calling from bam files
6. callpeaks.r / callpeaksWrapper.sh - to call ChIP peaks

## ChIPanalyser analysis
A combination of r and shell scripts were used for the ChIPanalyser analysis as follows: 

1. generateTable.r - to store all the necessary parameters for the analysis for each TF into one table
2. ChIPanalPerformAnalysis.R - to extract the parameters from the table and parse them, finally passing them to the performAnalysis.r function
3. performAnalysis.r - to perform the analysis; this can be called manually for indivisual TFs, or by the ChIpanalPerformAnalysis.sh script
